rtrlib/pfx/trie/trie-pfx.c

```diff
@@ -133,18 +153,31 @@ inline struct trie_node *pfx_table_get_root(const struct pfx_table *pfx_table, c
 
 int pfx_table_del_elem(struct node_data *data, const unsigned int index)
 {
+    struct data_elem *tmp;
+    struct data_elem deleted_elem = data->ary[index];
+
     //if index is not the last elem in the list, move all other elems backwards in the array
     if(index != data->len - 1) {
         for(unsigned int i = index; i < data->len - 1; i++) {
             data->ary[i] = data->ary[i+1];
         }
     }
+
     data->len--;
-    data->ary = realloc(data->ary, sizeof(struct data_elem) * data->len);
-    if(data->len != 0 && data->ary == NULL)
-        return PFX_ERROR;
-    else if(data->len == 0)
+    if (!data->len) {
+        free(data->ary);
         data->ary = NULL;
+        return PFX_SUCCESS;
+    }
+
+    tmp = realloc(data->ary, sizeof(struct data_elem) * data->len);
+    if (!tmp) {
+        data->ary[data->len] = deleted_elem;
```

Comment:
It is, except when the element that should be deleted is the last element in the list. In that case it would be redundant. I think the possible redundant operation doesn't really matter, that realloc() fails is very unlikely already. So I didn't add another if-statement for the case, to keep it more simple instead.

If the element that should be deleted isn't at the end of the list, all elements in the list gets shifted to the left one position.
The data->ary[index] is oververwritten with the element at data->ary[index + 1].
The element at data->ary[data->len - 2] is therefore the same than the one at data->ary[data->len - 1]. 
So I saved the data->ary[index] elem and insert it at data->ary[data->len - 1] if realloc() failed.